File: C,C++/p10530.cpp
#include <iostream>
#include <cmath>
#include <cstring>

using namespace std;

int main() {
    while (true) {
        int high = 10, low = 1, num;
        string tooHigh = "too high";
        string rightOn = "right on";
        cin >> num;
        if (num == 0) break;
        string command;
        getline(cin, command);
        getline(cin, command);
        while (command != rightOn) {
            if (command == tooHigh) {
                high = min(high, num-1); 
            } else {
                low = max(low, num+1);
            }

            cin >> num;
            getline(cin, command);
            getline(cin, command);
        }

        if (num >= low && num <= high) {
            cout << "Stan may be honest" << endl;
        } else {
            cout << "Stan is dishonest" << endl;
        }
    }

    return 0;
}

File: Java/p10530.java

import java.util.Scanner;

/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */

/**
 *
 * @author DarKnight98
 */
public class p10530 {
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        boolean lie = false;
        while (true) {
            int num, max = 10, min = 1;
            num = input.nextInt();
            if (num == 0)break;
            String command = input.nextLine();
            command = input.nextLine();
            while (!command.equals("right on")) {
                if (command.equals("too high")) {
                  max = Math.min(max, num - 1);
                } else {
                  min = Math.max(min, num + 1);
                }
                num = Integer.parseInt(input.nextLine().trim());
                command = input.nextLine().trim();
              }

              if (num >= min && num <= max) {
                System.out.println("Stan may be honest");
              } else {
                System.out.println("Stan is dishonest");
              }

        }
    }
}
